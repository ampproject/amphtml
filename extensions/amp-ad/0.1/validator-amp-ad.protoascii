#
# Copyright 2016 The AMP HTML Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the license.

# In the AMP Runtime, amp-ad and amp-embed are implemented in the samei
# extension, so we include them both here.

tags: {  # amp-ad
  html_format: AMP  # Ads are not allowed inside ads
  tag_name: "SCRIPT"
  # Typically the extension_spec takes care of this, but we have
  # a unique situation where we need to refer to this earlier in
  # the rules processing for the also_requires_tag_warning below.
  spec_name: "amp-ad extension .js script"
  extension_spec: {
    name: "amp-ad"
    allowed_versions: "0.1"
    allowed_versions: "latest"
    deprecated_recommends_usage_of_tag: "amp-embed"
    deprecated_recommends_usage_of_tag: "amp-ad"
    deprecated_allow_duplicates: true
  }
  attr_lists: "common-extension-attrs"
}
tags: {  # <amp-ad>
  html_format: AMP  # Ads are not allowed inside ads
  tag_name: "AMP-AD"
  # Typically we'd require the extension j/s, but for historical reasons
  # many pages don't have it, so it ends up late loaded and we warn, instead
  # of making this a validation error.
  also_requires_tag_warning: "amp-ad extension .js script"
  # If modifying disallowed_ancestors, then please also edit
  # extensions/amp-auto-ads/*/placement.js
  disallowed_ancestor: "AMP-APP-BANNER"
  disallowed_ancestor: "AMP-SIDEBAR"
  attrs: { name: "alt" }
  attrs: { name: "json" }
  attrs: {
    name: "src"
    value_url: {
      allowed_protocol: "https"
      allow_relative: true  # Will be set to false at a future date.
    }
    blacklisted_value_regex: "__amp_source_origin"
  }
  attrs: { name: "type" mandatory: true }
  attr_lists: "extended-amp-global"
  spec_url: "https://www.ampproject.org/docs/reference/components/amp-ad"
  amp_layout: {
    supported_layouts: FILL
    supported_layouts: FIXED
    supported_layouts: FIXED_HEIGHT
    supported_layouts: FLEX_ITEM
    supported_layouts: NODISPLAY
    supported_layouts: RESPONSIVE
  }
}
tags: {  # <amp-embed>
  html_format: AMP
  tag_name: "AMP-EMBED"
  # Typically we'd require the extension j/s, but for historical reasons
  # many pages don't have it, so it ends up late loaded and we warn, instead
  # of making this a validation error.
  also_requires_tag_warning: "amp-ad extension .js script"
  disallowed_ancestor: "AMP-APP-BANNER"
  disallowed_ancestor: "AMP-SIDEBAR"
  attrs: { name: "alt" }
  attrs: { name: "json" }
  attrs: {
    name: "src"
    value_url: {
      allowed_protocol: "https"
      allow_relative: true  # Will be set to false at a future date.
    }
    blacklisted_value_regex: "__amp_source_origin"
  }
  attrs: { name: "type" mandatory: true }
  attr_lists: "extended-amp-global"
  spec_url: "https://www.ampproject.org/docs/reference/components/amp-ad"
  amp_layout: {
    supported_layouts: FILL
    supported_layouts: FIXED
    supported_layouts: FIXED_HEIGHT
    supported_layouts: FLEX_ITEM
    supported_layouts: NODISPLAY
    supported_layouts: RESPONSIVE
  }
}
