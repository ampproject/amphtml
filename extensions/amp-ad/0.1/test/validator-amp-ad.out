FAIL
amp-ad/0.1/test/validator-amp-ad.html:45:4 The tag 'amp-ad with data-multi-size attribute' may not appear as a descendant of tag 'amp-fx-flying-carpet'. (see https://www.ampproject.org/docs/reference/components/amp-ad) [AMP_TAG_PROBLEM]
amp-ad/0.1/test/validator-amp-ad.html:49:4 The tag 'amp-embed with data-multi-size attribute' may not appear as a descendant of tag 'amp-fx-flying-carpet'. (see https://www.ampproject.org/docs/reference/components/amp-ad) [AMP_TAG_PROBLEM]
